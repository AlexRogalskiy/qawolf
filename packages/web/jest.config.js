const base = require("../../jest.config.base.js");

module.exports = {
  ...base,
  // tests use page.evaluate so we cannot collect coverage
  collectCoverage: false,
  name: "@qawolf/web",
  displayName: "@qawolf/web"
};
