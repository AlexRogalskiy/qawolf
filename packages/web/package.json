{
  "name": "@qawolf/web",
  "description": "qawolf web library",
  "version": "0.3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "build": "rollup --config rollup.ts",
    "prepublish": "npm run build"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "rollup": "^1.23.1",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-typescript": "^1.0.1"
  }
}
