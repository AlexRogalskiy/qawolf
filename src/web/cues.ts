import { getAttribute } from './attribute';
import { canTargetValue } from './element';
import { isDynamic } from './isDynamic';
import { SelectorPart } from './types';

const DEFAULT_ATTRIBUTE =
  'data-cy,data-e2e,data-qa,data-test,data-testid,/^qa-.*/';

// make sure to update CueTypeRank if editing this
const CSS_ATTRIBUTES = [
  'alt',
  'aria-label',
  'contenteditable',
  'for',
  'href',
  'name',
  'placeholder',
  'src',
  'title',
  'value',
] as const;

const CueTypes = [
  ...CSS_ATTRIBUTES,
  'attribute',
  'class',
  'id',
  'tag',
  'text',
] as const;

export type CueType = typeof CueTypes[number];

export type Cue = {
  level: number; // 0 is target, 1 is parent, etc.
  type: CueType;
  value: string;
};

type BuildAttributeCues = {
  attributes: string[];
  element: HTMLElement;
  level: number;
  useAttributeName?: boolean;
};

export type BuildCues = {
  attribute?: string;
  isClick: boolean;
  target: HTMLElement;
};

type BuildCuesForElement = {
  attributes: string[];
  element: HTMLElement;
  isClick: boolean;
  level: number;
};

type BuildTextCues = {
  element: HTMLElement;
  isClick: boolean;
  level: number;
};

export const buildAttributeCues = ({
  attributes,
  element,
  level,
  useAttributeName,
}: BuildAttributeCues): Cue[] => {
  const cues: Cue[] = [];

  attributes.forEach((attribute) => {
    const attributeValuePair = getAttribute({ attribute, element });
    if (!attributeValuePair) return;

    const { name, value } = attributeValuePair;
    const type = (useAttributeName ? name : 'attribute') as CueType;

    cues.push({ level, type, value: `[${name}="${value}"]` });
  });

  return cues;
};

export const buildCueValueForTag = (element: HTMLElement): string => {
  const tagName = element.tagName.toLowerCase();
  if (!element.parentElement) return tagName;

  const siblings = element.parentElement.children;
  const sameTagSiblings: HTMLElement[] = [];

  for (const sibling of siblings) {
    if (sibling.tagName.toLowerCase() === tagName) {
      sameTagSiblings.push(sibling as HTMLElement);
    }
  }

  if (sameTagSiblings.length < 2) {
    return tagName;
  }

  const nthIndex = sameTagSiblings.indexOf(element) + 1;
  if (nthIndex === 1) return tagName;

  return `${tagName}:nth-of-type(${nthIndex})`;
};

export const buildTextCues = ({
  element,
  isClick,
  level,
}: BuildTextCues): Cue[] => {
  if (!isClick) return [];

  let text = element.innerText.trim();

  if (
    element instanceof HTMLInputElement &&
    ['button', 'submit'].includes(element.type)
  ) {
    text = element.value;
  }

  if (
    !text ||
    text.length > 200 ||
    text.match(/[\n\r\t]+/) ||
    // ignore invisible characters which look like an empty string but have a length
    // https://www.w3resource.com/javascript-exercises/javascript-string-exercise-32.php
    // https://stackoverflow.com/a/21797208/230462
    text.match(/[^\x20-\x7E]/g)
  )
    return [];

  return [{ level, type: 'text', value: text }];
};

export const buildCuesForElement = ({
  attributes,
  element,
  isClick,
  level,
}: BuildCuesForElement): Cue[] => {
  let cssAttributes = [...CSS_ATTRIBUTES];

  if (!canTargetValue(element)) {
    cssAttributes = cssAttributes.filter((attribute) => attribute !== 'value');
  }

  const cues: Cue[] = [
    ...buildAttributeCues({ attributes, element, level }),
    ...buildAttributeCues({
      attributes: cssAttributes,
      element,
      level,
      useAttributeName: true,
    }),
    ...buildTextCues({ element, isClick, level }),
  ];

  element.classList.forEach((c) => {
    if (isDynamic(c)) return;

    cues.push({ level, type: 'class', value: `.${c}` });
  });

  if (element.id && !isDynamic(element.id)) {
    cues.push({ level, type: 'id', value: `#${element.id}` });
  }

  cues.push({ level, type: 'tag', value: buildCueValueForTag(element) });

  return cues;
};

export const buildCues = ({ attribute, isClick, target }: BuildCues): Cue[] => {
  const attributes = (attribute || DEFAULT_ATTRIBUTE).split(',');

  const cues: Cue[] = [];
  let element: HTMLElement = target;
  let level = 0;

  while (element) {
    cues.push(...buildCuesForElement({ attributes, element, isClick, level }));

    element = element.parentElement;
    level += 1;
  }

  return cues;
};

export const buildSelectorParts = (cues: Cue[]): SelectorPart[] => {
  const levels = [...new Set(cues.map((cue) => cue.level))];
  levels.sort().reverse();

  const parts: SelectorPart[] = [];

  levels.forEach((level) => {
    const cuesForLevel = cues.filter((cue) => cue.level === level);

    const textCues = cuesForLevel.filter((cue) => cue.type === 'text');
    if (textCues.length) {
      parts.push({ name: 'text', body: textCues[0].value });
      return;
    }

    cuesForLevel.sort((a, b) => {
      if (a.type === 'tag') return -1;
      if (b.type === 'tag') return 1;
      return 0;
    });

    const bodyValues = cuesForLevel.map((cue) => cue.value);

    parts.push({ name: 'css', body: bodyValues.join('') });
  });

  return parts;
};
