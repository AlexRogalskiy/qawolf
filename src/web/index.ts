import * as actions from "./actions";
import * as locator from "./locator";
import * as rank from "./rank";
import { runStep } from "./runStep";
import * as xpath from "./xpath";

const qawolf = {
  actions,
  locator,
  rank,
  runStep,
  xpath
};

export type QAWolf = typeof qawolf;

if (typeof window !== "undefined" && !(window as any).qawolf) {
  (window as any).qawolf = qawolf;
  console.log("loaded qawolf");
}
