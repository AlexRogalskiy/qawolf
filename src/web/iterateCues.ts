import { Cue } from './cues';

type CombineCuesOptions = {
  minPenalty: number;
  maxPenalty: number;
  startingGroup?: Cue[];
  startingGroupIndexes?: number[];
};

// A larger number for INCREMENT results in less unnecessary looping
// but potentially means that a single high-penalty attribute
// will be used rather than two low-penalty attributes
const MAX_PENALTY_INCREMENT = 5;
const MAX_PENALTY_BEFORE_FALLBACK = 100;

/**
 * @summary Sorts a list of Cues by their `penalty` value, lowest penalty first.
 */
const sortCues = (cues: Cue[]): Cue[] =>
  cues.sort((a, b) => {
    // put the lower penalty at the top
    if (a.penalty < b.penalty) return -1;
    if (a.penalty > b.penalty) return 1;

    return 0;
  });

/**
 * @summary Recursive generator function that provides gradually longer
 *   cue groups within a set penalty range. The returned cue array always
 *   includes at least one cue from level 0 because the correct target
 *   can never be selected without a level 0 cue. It also never includes
 *   the same cue twice because this would not add any specificity to a
 *   selector.
 *
 *   By calling this function in a loop with increasing penalty ranges,
 *   we are able to prioritize having a low total penalty, even if that
 *   penalty is the result of multiple cues.
 */
function* combineCues(allCues: Cue[], {
  minPenalty,
  maxPenalty,
  startingGroup = [],
  startingGroupIndexes = []
}: CombineCuesOptions): Generator<Cue[], void, undefined> {
  const trialGroups = [];
  const trialGroupsIndexes = [];

  for (let index = 0; index < allCues.length; index++) {
    // There is never any reason to add a cue twice to the same group
    if (startingGroupIndexes.includes(index)) continue;

    const cue = allCues[index];
    const expandedGroup = [...startingGroup, cue];

    // Groups that have no level=0 cues are useless
    const level0Cue = expandedGroup.find((cue) => cue.level === 0);
    if (!level0Cue) continue;

    // Groups that are outside the current acceptable penalty
    // range will be tried later or were already tried.
    const totalPenalty = expandedGroup.reduce((acc, cue) => acc + cue.penalty, 0);
    if (totalPenalty > maxPenalty || totalPenalty < minPenalty) continue;

    trialGroups.push(expandedGroup);
    trialGroupsIndexes.push([...startingGroupIndexes, index]);
  }

  yield* trialGroups;

  for (let index = 0; index < trialGroups.length; index++) {
    const trialGroup = trialGroups[index];
    const trialGroupIndexes = trialGroupsIndexes[index];
    yield* combineCues(allCues, {
      minPenalty,
      maxPenalty,
      startingGroup: trialGroup,
      startingGroupIndexes: trialGroupIndexes,
    });
  }
}

export function* iterateCues(cues: Cue[]): Generator<Cue[], void, undefined> {
  const cuesSortedByPenalty = sortCues(cues);

  let minPenalty = 0;
  let maxPenalty = 0;
  do {
    yield* combineCues(cuesSortedByPenalty, { minPenalty, maxPenalty });
    minPenalty = maxPenalty + 1;
    maxPenalty = maxPenalty + MAX_PENALTY_INCREMENT;
  } while (maxPenalty <= MAX_PENALTY_BEFORE_FALLBACK);
}
