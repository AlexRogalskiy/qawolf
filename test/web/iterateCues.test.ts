import { Cue, getCueTypesConfig } from '../../src/web/cues';
import { iterateCues } from '../../src/web/iterateCues';

const cueTypes: string[] = [
  'aria-label',
  'data-qa',
  'data-testid',
  'id',
  'text',
];

const buildCues = (level = 0): Cue[] => {
  const cueTypesConfig = getCueTypesConfig(['data-qa', 'data-testid']);
  return cueTypes.map((type) => {
    const { penalty } = cueTypesConfig[type];
    return {
      level,
      penalty,
      type,
      value: 'value'
    };
  });
};

describe('iterate cues', () => {
  it('orders attributes cues first then types with equal or higher rank', () => {
    const cues = [...buildCues(0), ...buildCues(1)];
    const iterated = Array.from(iterateCues(cues));
    expect(
      iterated.map((cues) => cues.map((c) => `${c.level}${c.type}`)),
    ).toMatchSnapshot();
  });
});
