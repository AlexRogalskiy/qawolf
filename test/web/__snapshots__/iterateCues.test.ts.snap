// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iterate cues orders attributes cues first then types with equal or higher rank 1`] = `
Array [
  Array [
    "0data-qa",
  ],
  Array [
    "0data-testid",
  ],
  Array [
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0data-qa",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0data-qa",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0data-qa",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0data-qa",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0data-qa",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0data-qa",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0data-testid",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0data-testid",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0data-testid",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0data-testid",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0data-testid",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0data-testid",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
  ],
  Array [
    "0id",
    "0data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-qa",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0id",
    "0data-testid",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-qa",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0id",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-testid",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0id",
    "1data-testid",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
  ],
  Array [
    "0text",
    "0data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-qa",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0text",
    "0data-testid",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-qa",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0text",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-testid",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0text",
    "1data-testid",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
  ],
  Array [
    "0aria-label",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-qa",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "0data-qa",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "0data-qa",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "0data-testid",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-testid",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-testid",
    "0data-qa",
    "1data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "0data-testid",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-qa",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-qa",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-qa",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-testid",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-testid",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-testid",
    "0data-qa",
    "1data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "0data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "1data-qa",
    "0data-qa",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "1data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "1data-qa",
    "0data-qa",
    "0data-testid",
  ],
  Array [
    "0aria-label",
    "1data-testid",
    "1data-qa",
    "0data-testid",
    "0data-qa",
  ],
]
`;
